nixos/modules/programs/singularity.nix

```diff
@@ -3,32 +3,66 @@
 with lib;
 let
   cfg = config.programs.singularity;
-  singularity = pkgs.singularity.overrideAttrs (attrs : {
-    installPhase = attrs.installPhase + ''
-      mv $out/libexec/singularity/bin/starter-suid $out/libexec/singularity/bin/starter-suid.orig
-      ln -s /run/wrappers/bin/singularity-suid $out/libexec/singularity/bin/starter-suid
-    '';
-  });
-in {
+in
+{
   options.programs.singularity = {
-    enable = mkEnableOption "Singularity";
+    enable = mkEnableOption "singularity" // {
+      description = ''
+        Whether to install Singularity/Apptainer
+        and add SUID support.
+      '';
+    };
+    package = mkOption {
+      default = pkgs.singularity;
+      defaultText = "pkgs.singularity";
+      description = ''
+        Singularity package to override and install. Typical values are
+        pkgs.singularity, pkgs.singularity-legacy, pkgs.apptainer, and pkgs.singularity-ce.
+      '';
+      type = types.package;
+    };
+    packageOverriden = mkOption {
+      default = mkDefault null;
+      defaultText = "lib.mkDefault null";
+      description = ''
+        The overriden singularity package to install.
+        <note>
+          <para>
+            This option is handled internally to place the overriden result of
+            <varname>programs.singularity.package</varname> with SUID support added.
+          </para>
+          <para>
+            Specify with <varname>lib.mkForce</varname> if you want to do it yourself.
+          </para>
+        </note>
+      '';
+      type = types.nullOr types.package;
+    };
   };
 
   config = mkIf cfg.enable {
-      environment.systemPackages = [ singularity ];
-      security.wrappers.singularity-suid =
-      { setuid = true;
+    programs.singularity.packageOverriden = (cfg.package.override {
```

Comment:
I'm also wondering if a full rebuild with `overrideAttrs` is actually preferable to, say, `symlinkJoin` with `postBuild`.
Note that the overridden package won't be built and cached by hydra, the end users will have to run the builds on their own machines.

How do other suid modules do this?

Reply:
On the other hand, I really think that with a NixOS module singularity _must_ know to look in `/run/opengl-driver/lib` (cf. the [proof of concept](https://gist.github.com/SomeoneSerge/cf1210bf2a3eb2d7ad8adb4dac1816e3) again)

Reply:
@SomeoneSerge Does this implementation breaks when `/run/opengl-driver/lib` doesn't exist? Or would there ever be such a case for NixOS?

We could add it into the `programs.singularity` module.

BTW, as Apptainer has deprecated the `* path` configuration directives, we'll need to inject `ldconfig` through the `defaultPath` in the upstream source code.

Reply:
In the mean time, as Apptainer has deprecated the `* path` configuration directives, we'll need to inject `ldconfig` through the `defaultPath` in the upstream source code.

I just add the `defaultPathInputs` attribute to make the overriding easier:

```diff
-    postInstall = (a.postInstall or "") + ''
-      sed -i 's|.*ldconfig path =.*|ldconfig path = ${ldconfig}|' $out/etc/${projectName}/${projectName}.conf
-    '';
+    defaultPathInputs = (a.defaultPathInputs or [ ]) ++ [ ldconfig ];
```

Reply:
> Does this implementation breaks when `/run/opengl-driver/lib` doesn't exist?

Hi. It shouldn't, but also the path exists on NixOS as long as `hardware.opengl.enable = true`